File: packages/cli/src/commands/deprecate.ts
import { Command } from 'commander';
import { icons, c, log, logJson, logError, withSpinner, getCurrentMode } from '../lib/output.js';
import { loadConfig } from '../lib/config.js';
import { createApiClient, ApiClientError } from '../lib/api-client.js';

export const registerDeprecateCommand = (program: Command): void => {
  program
    .command('deprecate <name>')
    .description('Deprecate a skill (or undo deprecation with --undo)')
    .option('--message <msg>', 'Deprecation message shown to users')
    .option('--undo', 'Remove deprecation')
    .action(async (name: string, opts: { message?: string; undo?: boolean }) => {
      const mode = getCurrentMode();

      // 1. Check auth
      const config = loadConfig();
      if (!config.token) {
        logError(
          'Not authenticated',
          'You must be logged in to deprecate skills.',
          `Run ${c.cmd('spm login')} to authenticate.`,
        );
        process.exitCode = 1;
        return;
      }

      const api = createApiClient(config);

      // 2. Build request body
      const isUndo = opts.undo ?? false;
      const body: Record<string, unknown> = isUndo
        ? { deprecated: false }
        : { deprecated: true, deprecated_msg: opts.message ?? 'This skill has been deprecated.' };

      // 3. Call API
      try {
        await withSpinner(
          isUndo ? `Removing deprecation for ${name}...` : `Deprecating ${name}...`,
          () => api.updateSkill(name, body),
        );

        if (isUndo) {
          log(`${icons.success} Un-deprecated ${c.name(name)}`);
        } else {
          log(`${icons.success} Deprecated ${c.name(name)}`);
          if (opts.message) {
            log(`  Message: ${c.dim(opts.message)}`);
          }
        }

        if (mode === 'json') {
          logJson({
            command: 'deprecate',
            status: 'success',
            name,
            deprecated: !isUndo,
            message: isUndo ? undefined : (opts.message ?? 'This skill has been deprecated.'),
          });
        }
      } catch (err) {
        if (err instanceof ApiClientError) {
          if (err.status === 404) {
            logError(
              'Skill not found',
              `${name} does not exist in the registry.`,
              `Run ${c.cmd('spm search')} to find the correct name.`,
            );
          } else if (err.status === 401) {
            logError(
              'Authentication failed',
              'Your token may have expired.',
              `Run ${c.cmd('spm login')} to re-authenticate.`,
            );
          } else if (err.status === 403) {
            logError('Permission denied', 'You can only deprecate skills you own.');
          } else {
            logError(
              isUndo ? 'Un-deprecate failed' : 'Deprecate failed',
              err.apiError.message,
              err.apiError.suggestion,
            );
          }
        } else {
          logError(
            isUndo ? 'Un-deprecate failed' : 'Deprecate failed',
            err instanceof Error ? err.message : String(err),
          );
        }
        process.exitCode = 1;
      }
    });
};

File: cli-go/cmd/deprecate.go
package cmd

import (
	"fmt"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var (
	deprecateMessage string
	deprecateUndo    bool
)

var deprecateCmd = &cobra.Command{
	Use:   "deprecate <@scope/name>",
	Short: "Deprecate a skill (or undo with --undo)",
	Long:  "Mark a skill as deprecated with an optional message. Use --undo to remove deprecation.",
	Args:  cobra.ExactArgs(1),
	RunE:  runDeprecate,
}

func init() {
	deprecateCmd.Flags().StringVar(&deprecateMessage, "message", "", "Deprecation message shown to users")
	deprecateCmd.Flags().BoolVar(&deprecateUndo, "undo", false, "Remove deprecation")
	rootCmd.AddCommand(deprecateCmd)
}

func runDeprecate(_ *cobra.Command, args []string) error {
	name := args[0]

	// 1. Check auth
	if Cfg.Token == "" {
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(map[string]any{
				"command": "deprecate",
				"status":  "error",
				"error":   "not_logged_in",
			})
		}
		Out.LogError("Not authenticated. Run %s to log in.", output.Cyan("spm login"))
		return fmt.Errorf("not logged in")
	}

	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)

	// 2. Determine deprecate vs undo
	deprecated := !deprecateUndo
	msg := deprecateMessage
	if !deprecated {
		msg = ""
	} else if msg == "" {
		msg = "This skill has been deprecated."
	}

	action := "Deprecating"
	if !deprecated {
		action = "Removing deprecation for"
	}

	sp := Out.StartSpinner(fmt.Sprintf("%s %s...", action, name))
	resp, err := client.Deprecate(name, deprecated, msg)
	output.StopSpinner(sp)

	if err != nil {
		var apiErr *api.APIError
		if isAPIError(err, &apiErr) {
			if Out.Mode == output.ModeJSON {
				return Out.LogJSON(map[string]any{
					"command": "deprecate",
					"status":  "error",
					"error":   apiErr.Code,
					"message": apiErr.Message,
				})
			}
			if apiErr.IsNotFound() {
				Out.LogError("Skill %s does not exist in the registry.", name)
				scopeHint(name)
			} else if apiErr.IsUnauthorized() {
				Out.LogError("Authentication failed. Run %s to re-authenticate.", output.Cyan("spm login"))
			} else {
				Out.LogError("%s failed: %s", action, apiErr.Message)
			}
			return fmt.Errorf("%s failed: %s", action, apiErr.Message)
		}
		Out.LogError("%s failed: %s", action, err)
		return fmt.Errorf("%s failed: %w", action, err)
	}

	if Out.Mode == output.ModeJSON {
		result := map[string]any{
			"command":    "deprecate",
			"status":     "success",
			"name":       resp.Name,
			"deprecated": resp.Deprecated,
			"updated_at": resp.UpdatedAt,
		}
		if resp.DeprecatedMsg != "" {
			result["message"] = resp.DeprecatedMsg
		}
		return Out.LogJSON(result)
	}

	if deprecated {
		Out.Log("%s Deprecated %s", output.Green(output.Icons["success"]), output.Cyan(name))
		if msg != "" {
			Out.Log("  Message: %s", output.Dim(msg))
		}
	} else {
		Out.Log("%s Un-deprecated %s", output.Green(output.Icons["success"]), output.Cyan(name))
	}

	return nil
}
